fem/bilininteg.hpp

```diff
@@ -3293,7 +3293,7 @@ class DGDiffusionIntegrator : public BilinearFormIntegrator
    void AddMultPAFaceNormalDerivatives(const Vector &x, const Vector &dxdn,
                                        Vector &y, Vector &dydn) const override;
 
-   const IntegrationRule &GetRule(int order, FaceElementTransformations &T);
+   const IntegrationRule &GetRule(int order, Geometry::Type geom);
```

Comment:
I think it is better to not change this -- keep it consistent with other `GetRule` methods.

Reply:
Should we implement `Mesh::GetTypicalFaceElementTransformations` or something similar?

The other one of this form is `DGTraceIntegrator::GetRule`. Looking at the PA code for that integrator, I see there is actually still a call to `GetFaceElementTransformations(0)`, so that code should be changed as well.

(The reason I made this change is because something like `GetTypicalFaceElementTransformations` was a little tricky).

Reply:
It seems like defining `GetTypical*` for transformations is taking things too far. Instead of constructing the transformation, it is simpler to call `GetRule` conditionally if there are actual elements in the local mesh and otherwise use an empty `IntegrationRule`. Was `Mesh::GetTypicalElementTransformation` need just to use it in a call to `GetRule`?

Reply:
`GetTypicalElementTransformation` is needed mostly (only?) for getting the quadrature rules.

I think it's better if the "user code" to get the quadrature rules doesn't have conditionals based on empty partitions. If the user needs to branch based on `ne == 0`, it's very easy to forget to do this, and then the bug is only caught if the code is actually run with empty partitions.

We could also change the `GetRule` interface (or, for backwards compatibility, add a new interface and deprecate the old one) that doesn't take an `ElementTransformation`, and instead takes some other information (geometry, order of the Jacobian determinant, etc.) that fulfills the same purpose.

Reply:
I'm beginning to think that the whole thing with the `GetTypical*` methods (that I started) is probably not the best approach to handle empty partitions. These calls to `GetTypical*` are only used because methods assume the a single type of element and order are present in the mesh. Thinking ahead, we want to generalize this to mixed meshes and variable orders, so these `GetTypical*` methods are just temporary addition that will become unnecessary in the future. Maybe we should try to deal with empty partitions in the context of dealing with mixed meshes and variable orders? What do you think?

Reply:
Yes, I think that the "GetTypical" approach is definitely a temporary solution that will not be necessary in the future.

I imagine that in the future, there will be some division of the elements into "classes" based on geometry, finite element polynomial degree, etc., and then all the elements of a certain class will be treated together in a batch. The code will have to handle potentially arbitrary numbers of classes, and so this will treat the empty partition case automatically (since in that case, there will just be zero classes).

Implementing that is probably a significant effort, and I don't know if there are concrete plans to do it currently. I think it's probably better to have a temporary "bandaid" fix for the empty partitions for now, and then refactor when the more general solution is implemented. Maybe we can mark the new "GetTypical" APIs as "internal" somehow, so that users know not to depend on them long-term.

Reply:
Okay sounds reasonable. As a solution of the above change in signature of `GetRule` -- do you mind keeping the original version (for consistency, backward compatibility, and maybe future use) and just add the new version as a separate overload?

Reply:
Done in d8df6b802045a8149c02a803429f79dee56dc348